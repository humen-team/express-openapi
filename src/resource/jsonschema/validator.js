import { isPlainObject, mapValues } from 'lodash';
import { Validator } from 'jsonschema';

import Reference from '../reference';
import { userIdFormat, uuidFormat } from './formats';

/* Replace `Reference` instances in the schema with an concrete path.
 *
 * OpenAPI 2.0 and 3.0.0 expect different root paths for resources. We avoid committing
 * to one path or another by using a `Reference` object instance and deferring spec building
 * until as late as possible.
 *
 * However, internal references may be needed for resource validation, so we replace
 * `Reference` instances with a _generic_ path during validation.
 */
function dereference(object, validator, path = '') {
    // recurse into lists
    if (Array.isArray(object)) {
        return object.map(
            (item) => dereference(item, validator, `${path}[]`),
        );
    }

    // recurse into objects
    if (isPlainObject(object)) {
        return mapValues(
            object,
            (value, key) => {
                if (key !== '$ref') {
                    // not a ref
                    return dereference(value, validator, `${path}.${key}`);
                }
                if (!(value instanceof Reference)) {
                    // not a valid ref
                    throw new Error('Expected $ref to be instance of Reference; did you use Resource.toRef()?');
                }
                // use '/foo' as the generic reference path
                const refId = value.id;

                // recurse to the referenced schema
                const { id, ...referenceSchema } = dereference(value.resource.schema, validator);

                // save the reference schema under the reference id *locally* within this validator.
                validator.addSchema({ id: refId, ...referenceSchema });

                // return the generic reference path in place of the Reference
                return refId;
            },
        );
    }

    // base case
    return object;
}

/* Create a JSON Schema validator for a specific resource.
 */
export default function createValidator(resource) {
    const validator = new Validator();

    validator.customFormats.uuid = uuidFormat;
    validator.customFormats['user-id'] = userIdFormat;

    const schema = dereference(resource.schema, validator);
    return {
        validate: (object) => validator.validate(object, schema),
    };
}
