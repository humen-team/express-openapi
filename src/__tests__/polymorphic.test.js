import request from 'supertest';

import { Namespace } from '..';
import { newApp } from './app';
import { Pet, PetType } from './polymorphic';
import validator, { openapi2, openapi3 } from './validate';

function create(pet) {
    return pet;
}

function search() {
    return {
        count: 2,
        items: [{
            additionalProperty: 'ignore',
            info: {
                lives: 9,
            },
            name: 'Felix',
            type: PetType.cat,
        }, {
            additionalProperty: 'ignore',
            info: {
                bestFriend: 'man',
            },
            name: 'Rex',
            type: PetType.dog,
        }],
    };
}

describe('polymorphic schema', () => {
    let app;

    beforeEach(() => {
        const pet = new Namespace('pet')
            .create({ input: Pet, output: Pet, route: create })
            .search({ output: Pet.toList(), route: search });

        app = newApp({ operations: [pet] });
    });

    describe('search', () => {
        it('returns polymorphic properties', async () => {
            const response = await request(app).get('/pet');
            expect(response.statusCode).toEqual(200);
            expect(response.body.count).toEqual(2);
            expect(response.body.items.length).toEqual(2);
            expect(response.body.items[0].additionalProperty).not.toBeDefined();
            expect(response.body.items[0].info).toEqual({ lives: 9 });
            expect(response.body.items[0].name).toEqual('Felix');
            expect(response.body.items[0].type).toEqual(PetType.cat);
            expect(response.body.items[1].info).toEqual({ bestFriend: 'man' });
            expect(response.body.items[1].name).toEqual('Rex');
            expect(response.body.items[1].type).toEqual(PetType.dog);
        });
    });

    describe('create', () => {
        it('supports well-formed cats', async () => {
            const response = await request(app).post('/pet').send({
                info: {
                    lives: 9,
                },
                name: 'Felix',
                type: PetType.cat,
            });
            expect(response.statusCode).toEqual(201);
        });
        it('validates malformed cats', async () => {
            const response = await request(app).post('/pet').send({
                info: {
                    lives: 'nine',
                },
                name: 'Felix',
                type: PetType.cat,
            });
            expect(response.statusCode).toEqual(422);
        });
        it('validates cats with additional properties', async () => {
            const response = await request(app).post('/pet').send({
                info: {
                    landsOnFeet: true,
                    lives: 9,
                },
                name: 'Felix',
                type: PetType.cat,
            });
            expect(response.statusCode).toEqual(422);
        });
        it('supports well-formed dogs', async () => {
            const response = await request(app).post('/pet').send({
                info: {
                    bestFriend: 'man',
                },
                name: 'Rex',
                type: PetType.dog,
            });
            expect(response.statusCode).toEqual(201);
        });
        it('validates malformed dogs', async () => {
            const response = await request(app).post('/pet').send({
                info: {
                    bestFriend: true,
                },
                name: 'Rex',
                type: PetType.dog,
            });
            expect(response.statusCode).toEqual(422);
        });
        it('validates cats with additional properties', async () => {
            const response = await request(app).post('/pet').send({
                info: {
                    chasesTail: true,
                    bestFriend: 'man',
                },
                name: 'Rex',
                type: PetType.dog,
            });
            expect(response.statusCode).toEqual(422);
        });
        it('validates unicorns', async () => {
            const response = await request(app).post('/pet').send({
                info: {
                    // we intentionally do not UnicornInfo
                },
                name: 'Amalthea',
                type: PetType.unicorn,
            });
            expect(response.statusCode).toEqual(422);
        });
    });

    describe('openapi 2.0', () => {
        it('is valid', async () => {
            const response = await request(app).get('/openapi/2.0');
            expect(response.statusCode).toEqual(200);
            validator.validate(response.body, openapi2, { throwError: true });
            expect(response.body).toMatchSnapshot();
        });
    });

    describe('openapi 3.0.0', () => {
        it('is valid', async () => {
            const response = await request(app).get('/openapi/3.0.0');
            expect(response.statusCode).toEqual(200);
            validator.validate(response.body, openapi3, { throwError: true });
            expect(response.body).toMatchSnapshot();
        });
    });
});
