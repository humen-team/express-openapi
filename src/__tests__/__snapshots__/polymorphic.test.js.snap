// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`polymorphic schema openapi 2.0 is valid 1`] = `
Object {
  "basePath": "/",
  "definitions": Object {
    "Error": Object {
      "properties": Object {
        "code": Object {
          "type": "integer",
          "x-nullable": false,
        },
        "message": Object {
          "type": "string",
          "x-nullable": false,
        },
        "status": Object {
          "type": "integer",
          "x-nullable": false,
        },
      },
      "type": "object",
    },
    "Pet": Object {
      "properties": Object {
        "info": Object {
          "$ref": "#/definitions/PetInfo",
        },
        "name": Object {
          "type": "string",
          "x-nullable": false,
        },
        "type": Object {
          "enum": Array [
            "cat",
            "dog",
            "unicorn",
          ],
          "type": "string",
          "x-nullable": false,
        },
      },
      "required": Array [
        "name",
        "type",
        "info",
      ],
      "type": "object",
    },
    "PetInfo": Object {
      "properties": Object {
        "bestFriend": Object {
          "type": "string",
          "x-nullable": false,
        },
        "lives": Object {
          "type": "integer",
          "x-nullable": false,
        },
      },
      "type": "object",
    },
    "PetList": Object {
      "properties": Object {
        "count": Object {
          "type": "integer",
          "x-nullable": false,
        },
        "items": Object {
          "items": Object {
            "$ref": "#/definitions/Pet",
          },
          "type": "array",
          "x-nullable": false,
        },
        "limit": Object {
          "type": "integer",
          "x-nullable": false,
        },
        "offset": Object {
          "type": "integer",
          "x-nullable": false,
        },
      },
      "required": Array [
        "items",
        "count",
      ],
      "type": "object",
    },
  },
  "info": Object {
    "description": "REST API",
    "title": "REST API",
    "version": "0.1.0",
  },
  "paths": Object {
    "/pet": Object {
      "get": Object {
        "operationId": "search",
        "parameters": Array [],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "200": Object {
            "description": "PetList as JSON",
            "schema": Object {
              "$ref": "#/definitions/PetList",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "pet",
        ],
      },
      "post": Object {
        "consumes": Array [
          "application/json",
        ],
        "operationId": "create",
        "parameters": Array [
          Object {
            "in": "body",
            "name": "pet",
            "required": true,
            "schema": Object {
              "$ref": "#/definitions/Pet",
            },
          },
        ],
        "produces": Array [
          "application/json",
        ],
        "responses": Object {
          "201": Object {
            "description": "Pet as JSON",
            "schema": Object {
              "$ref": "#/definitions/Pet",
            },
          },
          "default": Object {
            "description": "An error occurred",
            "schema": Object {
              "$ref": "#/definitions/Error",
            },
          },
        },
        "tags": Array [
          "pet",
        ],
      },
    },
  },
  "schemes": Array [
    "http",
  ],
  "swagger": "2.0",
}
`;

exports[`polymorphic schema openapi 3.0.0 is valid 1`] = `
Object {
  "components": Object {
    "schemas": Object {
      "Error": Object {
        "properties": Object {
          "code": Object {
            "nullable": false,
            "type": "integer",
          },
          "message": Object {
            "nullable": false,
            "type": "string",
          },
          "status": Object {
            "nullable": false,
            "type": "integer",
          },
        },
        "type": "object",
      },
      "Pet": Object {
        "properties": Object {
          "info": Object {
            "$ref": "#/components/schemas/PetInfo",
          },
          "name": Object {
            "nullable": false,
            "type": "string",
          },
          "type": Object {
            "enum": Array [
              "cat",
              "dog",
              "unicorn",
            ],
            "nullable": false,
            "type": "string",
          },
        },
        "required": Array [
          "name",
          "type",
          "info",
        ],
        "type": "object",
      },
      "PetInfo": Object {
        "properties": Object {
          "bestFriend": Object {
            "nullable": false,
            "type": "string",
          },
          "lives": Object {
            "nullable": false,
            "type": "integer",
          },
        },
        "type": "object",
      },
      "PetList": Object {
        "properties": Object {
          "count": Object {
            "nullable": false,
            "type": "integer",
          },
          "items": Object {
            "items": Object {
              "$ref": "#/components/schemas/Pet",
            },
            "nullable": false,
            "type": "array",
          },
          "limit": Object {
            "nullable": false,
            "type": "integer",
          },
          "offset": Object {
            "nullable": false,
            "type": "integer",
          },
        },
        "required": Array [
          "items",
          "count",
        ],
        "type": "object",
      },
    },
  },
  "info": Object {
    "description": "REST API",
    "title": "REST API",
    "version": "0.1.0",
  },
  "openapi": "3.0.0",
  "paths": Object {
    "/pet": Object {
      "get": Object {
        "operationId": "search",
        "parameters": Array [],
        "responses": Object {
          "200": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/PetList",
                },
              },
            },
            "description": "PetList as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "pet",
        ],
      },
      "post": Object {
        "operationId": "create",
        "parameters": Array [],
        "requestBody": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "$ref": "#/components/schemas/Pet",
              },
            },
          },
          "required": true,
        },
        "responses": Object {
          "201": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Pet",
                },
              },
            },
            "description": "Pet as JSON",
          },
          "default": Object {
            "content": Object {
              "application/json": Object {
                "schema": Object {
                  "$ref": "#/components/schemas/Error",
                },
              },
            },
            "description": "An error occurred",
          },
        },
        "tags": Array [
          "pet",
        ],
      },
    },
  },
  "servers": Array [
    Object {
      "url": "/",
    },
  ],
}
`;
